packages/java-parser/src/productions/expressions.js

```diff
@@ -653,10 +661,11 @@ function defineRules($, t) {
     }
 
     // in the middle of a "classReferenceType"
-    $.OPTION2(() => {
+    try {
```

Comment:
i don't get this

Reply:
Same as below, the OPTION was throwing an error, so I tried replacing it with this hack

Reply:
I just tested with keeping the original code, and it fails when trying to parse

```java
public class DispatcherServletsMappingDescriptionProvider {

	private static <T> List<T> describe() {
		for (HandlerMappingDescriptionProvider<?> descriptionProvider : descriptionProviders) {
			
		}
	}
}
```

with `compilationUnit` as entryPoint in the `single-sample-runner.js` script. I got this error:

```
/home/cdessoude/Documents/workplace/prettier-java/node_modules/chevrotain/lib/src/parse/parser/traits/recognizer_engine.js:188
            throw e;
            ^

TypeError: Cannot read property 'call' of undefined
    at JavaParser.RecognizerEngine.optionInternalLogic (/home/cdessoude/Documents/workplace/prettier-java/node_modules/chevrotain/lib/src/parse/parser/traits/recognizer_engine.js:227:27)
    at JavaParser.RecognizerEngine.optionInternal (/home/cdessoude/Documents/workplace/prettier-java/node_modules/chevrotain/lib/src/parse/parser/traits/recognizer_engine.js:196:25)
    at JavaParser.RecognizerApi.OPTION2 (/home/cdessoude/Documents/workplace/prettier-java/node_modules/chevrotain/lib/src/parse/parser/traits/recognizer_api.js:108:21)
    at JavaParser.$.RULE (/home/cdessoude/Documents/workplace/prettier-java/packages/java-parser/src/productions/expressions.js:654:7)
    at JavaParser.invokeRuleWithTry (/home/cdessoude/Documents/workplace/prettier-java/node_modules/chevrotain/lib/src/parse/parser/traits/recognizer_engine.js:119:33)
    at JavaParser.wrappedGrammarRule [as isRefTypeInMethodRef] (/home/cdessoude/Documents/workplace/prettier-java/node_modules/chevrotain/lib/src/parse/parser/traits/recognizer_engine.js:133:38)
    at JavaParser.BACKTRACK_LOOKAHEAD.ACTION (/home/cdessoude/Documents/workplace/prettier-java/packages/java-parser/src/parser.js:182:27)
    at JavaParser.RecognizerApi.ACTION (/home/cdessoude/Documents/workplace/prettier-java/node_modules/chevrotain/lib/src/parse/parser/traits/recognizer_api.js:21:21)
    at JavaParser.BACKTRACK_LOOKAHEAD (/home/cdessoude/Documents/workplace/prettier-java/packages/java-parser/src/parser.js:175:17)
    at JavaParser.$.RULE (/home/cdessoude/Documents/workplace/prettier-java/packages/java-parser/src/productions/expressions.js:314:35)
```

Reply:
maybe you need to only compute this after the recording phase has finished?

Reply:
you could move this calculation to the last step of the constructor for example

Reply:
You are talking about the `firstUnaryExpressionNotPlusMinus` issue, right (not about the `$.OPTION`) ? You would try to compute it after the this.performSelfAnalysis() in the JavaParser constructor ?

Reply:
Yes you are correct I must have replied in the wrong thread 😄